### ISSUE TITLE
kube-addon-update revamp

### ISSUE THREAD

**Issue by mikedanese** [kubernetes/kubernetes#23233]
kube-addon-update needs to support DaemonSets and Deployments which have different update strategies. It's half python and half shell so it can't run on coreos. It has no tests.

v2 should:
- support appropriate update strategies for DaemonSets and Deployments
- use kubectl apply
- be written entirely in python or go
- have tests
- run in a static pod on the master
- Address the current [limitations](https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kube-addons/kube-addon-update.sh#L25-L42)

ref #23174

**Comment by mikedanese** [comment]
I don't think the golang client is particularly useful here. What we really want to do is to script against kubectl which has patch/apply/replace logic builtin. We can exec or use kubectl as a library.

**Comment by mikedanese** [comment]
cc @roberthbailey @zmerlynn

**Comment by mikedanese** [comment]
This can probably live in the contrib/ repo.

**Comment by roberthbailey** [comment]
I thought we already had an issue to replace it but I don't see one. https://github.com/kubernetes/kubernetes/issues/7459 https://github.com/kubernetes/kubernetes/issues/9469 and https://github.com/kubernetes/kubernetes/issues/9849 all seem related. 

@zmerlynn has thought about this a bit and should definitely chime in.

**Comment by zmerlynn** [comment]
It actually does have a test: https://github.com/kubernetes/kubernetes/blob/4e29c580a8e274875f442be9d97a45baf1139aa8/test/e2e/addon_update.go

**Comment by zmerlynn** [comment]
It can't run on CoreOS for multiple reasons, which are all unrelated to this script. The fact that it's shell and python is *because* it's Salt-only.

If you're going to crack this nut for both CoreOS and Salt-based, the problem is that you need to flow the add-on configuration from somewhere. We've just started to have the right machinery for that, barely, but we still don't have a central authority for cluster config in the cluster. I'd recommend tabling that because it's just a swamp to wander into. But presuming we had a system that could tell you the desired add-on state for the currently running cluster, my goal was to eventually write a proper reconciler in Go that would get you there. (The shell script was a hack by @marekbiskup, on a super early hack by me, all for expediency.)

We can actually write the reconciler right now if you use the manifest directory approach and you use Salt as one annealer for that and then figure something else out for CoreOS. That part is really messy.

**Comment by mikedanese** [comment]
Even if it's not the complete solution, we will get a lot of mileage on systems like coreos by moving the update manager into a static pod(#9469). Instead of having to rewrite the entire addon management suite, they will only have to figure out distributing the addon configs.

**Comment by zmerlynn** [comment]
The actual pod-vs-daemon is mostly a distribution detail, to to be honest, and doesn't change what I was saying in https://github.com/kubernetes/kubernetes/issues/23233#issuecomment-199312020 (you can, in fact, distribute the existing script as a pod, even, and it doesn't change anything). You still have to get the addon config to the pod somehow. In a unicorns-and-rainbows world, it would just query something else for its config. For iterative development, though, I'd suggest the existing manifest approach.

**Comment by mikedanese** [comment]
ref #23471

**Comment by danehans** [comment]
ref https://github.com/kubernetes/contrib/issues/667

**Comment by rutsky** [comment]
@mikedanese can you ping me when this will be ready to try? I can try to support addon manager in contrib/ansible scripts.

**Comment by bgrant0607** [comment]
Also, please change from "kubectl stop" to "kubectl delete". stop has been deprecated for about a year.

**Comment by luxas** [comment]
> be written entirely in python or go

Please write in go

> run in a static pod on the master

:+1:

**Comment by mikedanese** [comment]
Before rewriting in go, we should evaluate wether helm can do what we need.

**Comment by luxas** [comment]
I have some ideas for addon-manager v2:

_The target addons here are cluster-internal ones that are inside Kubernetes core. User-made addons should be managed via Helm._

 - It should be written in Go and exec kubectl if the golang client isn't good enough to use for this
 - The addon-manager should live inside the cluster as a ReplicaSet.
   - It could be bootstrapped from a static pod which creates the real ReplicaSet.

The addon-manager should bootstrap mandatory addons automatically:
 - dns
 - dashboard
 - maybe helm

Instead of using files from one master's filesystem (not good for HA, hard to update and maintain), configmaps should be use for storing the config. Everything should be inside the kube-system namespace. `addon.status.${addon_name}` should have a literal whether it should be enabled or not.
This makes it easy to toggle the addon state, which is hard today (remove or move a manifest on a fs).
The manifest itself should be placed in `addon[.manifest].${addon_name}`.

Here are the ~commands it would automatically run with kubectl:
```
kubectl --namespace kube-system create configmap addon.status.dns --from-literal enabled=true
kubectl --namespace kube-system create configmap addon.dns --from-file /addons/dns.yaml

kubectl --namespace kube-system create configmap addon.status.dashboard --from-literal enabled=true
kubectl --namespace kube-system create configmap addon.dashboard --from-file /addons/dashboard.yaml
```

The addon-manager should `WATCH` ConfigMaps in the kube-system namespace => no `while true` loops with `sleep`; just do things when config changes. When `status` changes, it'll react to that, and when the manifest changes, it should upgrade/replace.

All manifests that are in Kubernetes core should be in the container by default => if the admin wants to enable cluster monitoring, he only need to run this:
```
kubectl --namespace kube-system create configmap addon.status.cluster-monitoring --from-literal enabled=true
```
This makes it easy for third-party tools to quickly enable/disable addons.

Releasing:
1) The addon-manager should be released with every new Kubernetes version, and copy in the manifests into the image (plus image revision in the tag) or
 - Con: Have to be often released.
2) The addon-manager should detect cluster version from the apiserver and download default manifests for the given version from github 
 - Con: Internet dependency.

Recap:
We'll have an automatic "everything included" setup with no extra scripting dependency => easier deployments.
If the admin wants to deploy a specific dns addon (not the default one)
```
kubectl --namespace kube-system edit configmap addon.dashboard
# => and set the content to the new manifest
```
If the admin wants to disable the dashboard:
```
kubectl --namespace kube-system edit configmap addon.status.dashboard
# => and set enabled=false

# Via commandline:
kubectl --namespace kube-system get configmap addon.status.dns -o yaml | sed "s/enabled: \"true\"/enabled: \"false\"/g" | kubectl replace -f -
```

Here a `kubectl set configmap-(literal|file)` command would be very useful.

Let me know what you think.
@mikedanese @justinsb @jbeda

**Comment by gmarek** [comment]
cc @wojtek-t @mwielgus

**Comment by luxas** [comment]
cc @kubernetes/sig-cluster-lifecycle ^

**Comment by bgrant0607** [comment]
Thanks for the detailed proposal @luxas. 

If one could update ConfigMap objects, then one could update the addon objects. Perhaps the argument is that the latter types of updates would be more complex. However, the desired state of the addons should be represented by their API objects themselves. The source for the updated configurations just needs to come from somewhere. 

If there were an update client running somewhere, it could read the manifests (e.g., from the release tarball) and apply them. The addon manager does that on the master node in order to address the case where there isn't an update client running elsewhere, but then the files have to be copied to there somehow (e.g. with Salt).

I don't see why we're reading files from a host, though. I also don't see why we need a continuously running process to perform what should be one-time updates. Is there a doc, issue, or PR that has discussion about how we settled on the current approach?

Why not bundle the update script/program and addon manifests in a container and launch it as a Job upon upgrade? Each release could have a default container, but only one parameter (image name/tag) would be needed in order to change to a different one. If we wanted to make some addons optional or provide other configuration options, that information could be stored in a ConfigMap read by the Job.

**Comment by gmarek** [comment]
@bgrant0607 - it's really handy to have this running continuously when debugging add-ons. One can push a new version and update the manifest on a running cluster to get a dev version in and iterate faster. Whatever we decide to do I'd like to keep this particular flow easy enough (firing a job that will read manifests or something is good enough, but I don't want to make it complex).

**Comment by mikedanese** [comment]
cc @thockin @MrHohn

**Comment by bgrant0607** [comment]
Copied from #42756:

Relevant text from the layers doc (join kubernetes-dev to access):
https://docs.google.com/document/d/1XkjVm4bOeiVkj-Xt1LgoGiqWsBfNozJ51dyI-ljzt1o/edit#

We need a built-in add-on manager (not unlike static pod manifests, but at the cluster level) so that we can automatically add self-hosted components and dynamic configuration to the cluster, and so we can factor out functionality from existing components in running clusters. 

At its core would be a pull-based declarative reconciler, as provided by the current add-on manager and as described in the whitebox app management doc (also join kubernetes-dev to access):
https://docs.google.com/document/d/1S3l2F40LCwFKg6WG0srR6056IiZJBwDmDvzHWRffTWk/edit#

See the picture at the end.

I'd like the mechanism to be able to pull from at least HTTP, HTTPS (e.g., S3 or GCS), and git. 

As for how to configure where it should pull from -- that would be a namespaced API.

Where do the turtles for addons and self-hosted components end? My first instinct is component configuration, but it needs more thought, since I could easily imagine wanting to use this to pull component configuration ConfigMaps.

Where should this live? I'd say controller-manager, at least for now. 

cc @mtaufen @pwittrock @smarterclayton @philips @thockin @timstclair @frankgreco

**Comment by bgrant0607** [comment]
Obviously I agree with @gmarek now.

**Comment by colemickens** [comment]
As this is relevant to cluster lifecycle: **cc:** @kubernetes/sig-cluster-lifecycle-misc (didn't realize this had fragmented into four different GH teams, hope this one is relevant)

**Comment by bgrant0607** [comment]
@colemickens Please don't use the misc group unless there isn't a more relevant one. This is a feature request / proposal. @kubernetes/sig-cluster-lifecycle-feature-requests is notified above. @kubernetes/sig-cluster-lifecycle-proposals is notified now. :-)

**Comment by timstclair** [comment]
> As for how to configure where it should pull from -- that would be a namespaced API.

What would be the authenticated account when applying? At least for kube-system (addons) the account would need pretty wide privileges, probably root-on-cluster. That means that configuration becomes a very high-value target (and we should probably disallow HTTP), so the endpoint itself as well as the configuration would need to be carefully protected. /cc @kubernetes/sig-auth-feature-requests

**Comment by bgrant0607** [comment]
Why not Helm?

1. That would be a system layering violation, according to the layers in the proposal. 
2. Helm is imperative, not declarative.
3. Something has to start Tiller. Probably the addon manager. 
4. Does Tiller respect authorization policy yet?

I could probably think of more reasons.

**Comment by bgrant0607** [comment]
@timstclair Excellent question. Yes, I'd only expect HTTP to be used in dev clusters, such as minikube.

**Comment by bgrant0607** [comment]
The reconciler could be associated with a ServiceAccount for authorization purposes.

Credentials for private/protected endpoints or repos could be in a Secret. We punted on that in gitRepo volume sources. #831 would avoid a new mechanism by just running a container in a Pod, but I don't think we want that in this case.

We'd need another way to automatically create/populate the reconciler API and secret (if necessary) for kube-system, but at least that wouldn't need to be the full addon manager.

Somewhat off topic, but Openshift can instantiate a Template for each new [project](https://docs.openshift.org/latest/dev_guide/projects.html#dev-guide-projects) (think namespace):
* https://docs.openshift.org/latest/admin_guide/managing_projects.html#template-for-new-projects
* https://docs.openshift.org/latest/dev_guide/templates.html
* https://github.com/kubernetes/community/blob/master/contributors/design-proposals/templates.md

**Comment by bgrant0607** [comment]
From @timstclair:

Reasons to do so:
- consistency with Kubernetes production code (easier for devs to manage)
- robustness (bash isn't exactly known for being a robust production language)
- reduced base image (less CVE management overhead)
- can be built with release branch code (avoid out of date kubectl problems: https://github.com/kubernetes/kubernetes/issues/34943, https://github.com/kubernetes/kubernetes/issues/41817)

**Comment by bgrant0607** [comment]
Current code: https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/addon-manager

**Comment by bgrant0607** [comment]
The diagram from the app management doc can also be found here (join kubernetes-dev to access):

https://docs.google.com/presentation/d/17AqTn5QisaD2F3pz2buZUVW1fdlIW-Mv_zIiRBrTaSY/edit#slide=id.g1cac0faa55_0_15

**Comment by bgrant0607** [comment]
I'd be ok with focusing on the addon manager first and add the API later, provided we keep in mind that we'll want consistent semantics.

**Comment by justinsb** [comment]
@bgrant0607 we are working on this in sig-cluster-lifecycle.  I'll invite you to participate in our discussions going forwards.

Clarifying question though: what functionality does `kubectl apply -f <url> --prune` _not_ have?  i.e. what is in scope here?

As far as I can see, the additional functionality you are asking for:

* updating as the source document changes (automatically?)
* specifying the prune labels in the manifest
* additional data sources (s3, gcs)
* ability to run this server side in future
* API configuration if it is running server-side

Are there any other significant features?

In particular, I'm not clear whether you see templating as in-scope or out-of-scope?

By templating I mean templating that this layer is aware of.  If you launch a supervisor process which then performs some templating or tailoring (like `kube-dns-autoscaler`) of managed pods I would treat that as out-of-scope.  This of course has the advantage that we don't need to bikeshed on `{` vs `{{` vs `${` but can instead leave that up to the supervisor developer.

cc @errordeveloper

**Comment by bgrant0607** [comment]
@justinsb Is there a proposal from sig-cluster-lifecycle? I didn't find one other than this issue.

What do I envision beyond apply + your list?

Ensure exists was added recently: #40108. We should decide whether we want to keep it.

I could imagine adding support for the equivalent of kubectl create secret and create configmap (generating API resources from raw data). We'll want declarative support for that in other contexts also, though.

I do view templating as out of scope for this feature. I still want the [Template API](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/templates.md). And, yes, custom controllers like kube-dns-autoscaler would be fine.

**Comment by errordeveloper** [comment]
I think that key argument for moderately complex client-side substitution VS supervisor is about proving that most of the use-case don't need supervision throughout the lifetime of the resources in question. Most parameters I've encountered in our add ons only need to be set during installation.

Additionally, I think there is a problem with supervisors for the users, as lines get more blurred - who manages what?
With client-side templates, one can render the add-on and examine it before installing. They can even render and check-in, and install only once checked-in.

I've been working on [`kubegen`](https://github.com/errordeveloper/kubegen) for a little while now, which attempts to solve some of the key issues around re-usability. It also tries to flatten some of the deeply nested structures (e.g. `spec.template.spec`), but that is somewhat separate as far as this issue is concerned. With this in mind, I do believe that `kubegen` has [a clear line of what would be reasonable to have and what's overkill](https://github.com/errordeveloper/kubegen/tree/fc09d52f8f41acb660a87a8b255a8b2f38686dc4#motivation--high-level-goals).

**Comment by bgrant0607** [comment]
Related: 
#39712
#41817
#30329

**Comment by bgrant0607** [comment]
@errordeveloper Thanks for the pointer to kubegen. It's been on my list, but I haven't had time to look at it.

Anyway, my point was that I consider templating to be out of scope for the addon manager. Whether it would be done client-side or server-side is somewhat beside the point.

**Comment by bgrant0607** [comment]
kops addon type:
https://github.com/kubernetes/kops/blob/master/channels/pkg/api/channel.go

Example:
https://github.com/kubernetes/kops/blob/master/addons/kubernetes-dashboard/addon.yaml

Update checker/applier:
https://github.com/kubernetes/kops/blob/master/channels/cmd/channels/apply_channel.go

**Comment by TamalSaha** [comment]
How about using the Helm/Tiller engine and deploy a "chart" with all the addons? That will also solve updating addons later.

cc: @technosophos

**Comment by roberthbailey** [comment]
@TamalSaha - see @bgrant0607's [comment earlier in the thread](https://github.com/kubernetes/kubernetes/issues/23233#issuecomment-285524825) about why Helm/Tiller isn't a good fit.

**Comment by bgrant0607** [comment]
@justinsb proposal: https://docs.google.com/document/d/1Laov9RCOPIexxTMACG6Ffkko9sFMrrZ2ClWEecjYYVg/edit?ts=58f28822#

**Comment by piosz** [comment]
cc @fgrzadkowski

**Comment by resouer** [comment]
I think https://github.com/kubernetes/kubeadm/issues/247 also related to this, it will be fixed after addon can be updated easily.

**Comment by roberthbailey** [comment]
Bumping this from 1.7 to 1.8.

**Comment by bgrant0607** [comment]
We also need a way to shut down addons during cluster teardown. #4630 #10179 #16337

**Comment by kfox1111** [comment]
@bgrant0607 @technosophos 

> bgrant0607 commented on Mar 9
> Why not Helm?
>    That would be a system layering violation, according to the layers in the proposal.

I run tiller in a static pod on one of my production systems. works great. It even allows you to manage sdn in a helm chart.

>    Helm is imperative, not declarative.

I don't follow. Its declarative. Or are you referring to the lack of direct kubectl support.

If tiller supported a 3rd party api, would that meet the requirements?

User uploads values.yaml file as a 3rd party object?

>    Something has to start Tiller. Probably the addon manager.

Or it be made a core service ala kube-apiserver, kube-controller-manager, kube-scheduler.

Or, what if it was the addon manager entirely?

>    Does Tiller respect authorization policy yet?

It has some authentication options. Not sure what the concern is.

> I could probably think of more reasons.

I think helm should not be dismissed quickly. It is good enough for launching and managing a lot of apps. I've used it to deploy kube-dashboard: https://github.com/kubernetes/charts/tree/master/stable/kubernetes-dashboard

It can deploy kube-dns compatible system via https://github.com/kubernetes/charts/tree/master/stable/coredns

ingress with https://github.com/kubernetes/charts/tree/master/stable/nginx-ingress

I've even gotten a prototype of deploying an easily configurable flannel here: https://github.com/kfox1111/charts/tree/flannel/stable/flannel

So, lets come up with the gaps in why helm can't serve the role of addon-manager  and see if they are a small enough list that helm can't be enhanced to cover the use case. I suspect the list is actually quite small. If so, it might be much less effort to tweak helm to support the use case then reimplement much of the functionality in addon-manager.

**Comment by luxas** [comment]
cc @justinsb Do you have a doc for the new addons API yet?

**Comment by cheld** [comment]
@luxas FYI

in Dashboard we have requests to allow some kind of plugin extentions. Such extentions typically, need a backend to be deployed. e.g. Weave Scope https://github.com/kubernetes/dashboard/pull/1949

We had some internal discussion, if Dashboard should deploy such backend containers in case a user enables a plugin. However, I see some overlap with this proposal. 

Currently, we tend to not deploy anything. Just try to detect the backend and then enable the frontend integration. E.g. in other words, in case tiller is deployed in the right version, Dashboard will show Helm integration.

But it it might be interessting at some point to check if we could/should use this mechanism.

CC @bryk, @floreks

**Comment by bgrant0607** [comment]
Recap:

An "add on" is any Kubernetes API resource that is managed as part of the cluster lifecycle. In this issue, it specifically refers to the ones in the cluster/addons directory managed by the kube-up add-on manager script:
https://github.com/kubernetes/kubernetes/tree/master/cluster/addons
https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/addon-manager

An add-on could be a deployment such as kube-dns, daemonsets such as fluentd, policies such as RBAC rules, component configuration configmaps, etc. Since these are Kubernetes resources, they are all inherently cluster-level. Node-level behavior is achieved via DaemonSet.

Updating the set of problems to address:

* Implementation problems: /cluster is deprecated, we want to remove it from the kubernetes repo, and it will be removed from the release bundle. We're moving away from Salt, Jinja, and scripts. Some use Salt and Jinja, some (e.g., OS images without Salt and/or Python) use sed to replace the parameters and some other way to distribute the files. Scripts have subtle failure modes. Testing is hard. The image has security problems. Etc. In short, the add-on manager needs to be re-implemented and deployed using totally different mechanisms.
* Scope changes: add-ons are needed in multiple deployment tools, such as kops and minikube. More broadly, as we decompose the system into more and more components and as we move to self-hosting, component configuration via configmaps, policy resources, and so on, all clusters will need add-ons. It's not accurate to think of all addons as optional, though they may be pluggable and configurable.
* User configuration/customization: Users need to be able to configure add-on behavior, such as the cluster DNS subdomain. EnsureExists was added in order to support that without needing to fork the configuration and push it out with Salt, but is confusing. I'd like an apply-compatible approach that doesn't involve parameter substitution.
* Cluster-level and node-level resource management: As components use more resources and as we add more components, more cluster and node resources are required to run Kubernetes. Proliferation of many orthogonal, small add-ons wastes resources. Managing resources via Jinja and Salt is impractical. Add-on component resources and replica counts should be set separately from the shipped configurations, and/or they should be autoscaled. We need to protect add-ons from eviction by giving them higher scheduling/eviction priority than user applications, but that puts user applications at risk in the case of insufficient resources, and we don't have a way to ensure that enough resources have been provisioned prior to upgrading.
* Cluster component versioning (#4855): Add-ons versions are tied to Kubernetes release versions, and are typically upgraded as master instances are upgraded. This doesn't work in HA clusters. Also, there is a need to deploy different pods on nodes according to kubelet release version. One could label the nodes with the kubelet release version and instantiate multiple DaemonSets simultaneously, each with a nodeSelector targeting the desired kubelet version. This approach was essentially used to convert fluentd from a static pod to a DaemonSet, though only one DaemonSet was needed in that case. The cluster-level version-skew story is complicated by the use of alpha features in addons, and by relying on beta features, especially ones that don't exist in prior releases. This means that it's not safe to rollback the cluster until these resources are reverted, since the APIs may not exist in the reverted cluster, and means that it's not safe to roll forward, in general, either, since the features (alpha ones, anyway) may disappear. Examples: Deployment, ReplicaSet, and DaemonSet will have v1 APIs in 1.9, so we may remove the beta APIs in 1.10 or 1.11. That would break all existing add-on components that haven't been updated by then. I believe RBAC will go v1 in 1.8, so associated RBAC rules would break if not updated before the beta API is removed. At some point we'll remove the scheduler.alpha.kubernetes.io/critical-pod annotation, so any addons with that annotation that don't use its replacement will lose their eviction protection. Taints and labels used by cluster management tools such as kops and kubeadm change regularly, so addons with outdated selectors and tolerations may stop scheduling entirely.
* Authorization policy / protection: There is a desire to protect add-ons from accidental deletion and mutation (#10179), and, by some, to subdivide cluster administrator superpowers into cluster-management root and application-management root.

**Comment by piosz** [comment]
cc @x13n

**Comment by bgrant0607** [comment]
I proposed an application definition format and customization approach that I'd like to use with add-ons in my declarative application management doc:
https://goo.gl/T66ZcD

**Comment by bgrant0607** [comment]
Example of a node-level add-on rollout: https://github.com/kubernetes/kubernetes/pull/50705

**Comment by bgrant0607** [comment]
We also have a hodgepodge of auto-scaling mechanisms right now:
* https://github.com/thockin/cluster-proportional-vertical-autoscaler
* https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler
* https://github.com/kubernetes/autoscaler/tree/master/addon-resizer
* https://github.com/kubernetes-incubator/cluster-proportional-autoscaler

And some use operators for stateful addons:
* https://github.com/coreos/etcd-operator
* https://github.com/coreos/prometheus-operator
* https://github.com/upmc-enterprises/elasticsearch-operator

**Comment by bgrant0607** [comment]
Also related: #4630, #16337, #4855

**Comment by bgrant0607** [comment]
And #42840, #42917 for release-specific node-level addon rollouts.

**Comment by tamalsaha** [comment]
@bgrant0607, is this proposal worked on somewhere? How can anyone can be in the loop for this?

**Comment by bgrant0607** [comment]
Ref #51862 regarding resource protection

**Comment by bgrant0607** [comment]
@tamalsaha Not AFAIK. Activity should be tracked in https://github.com/kubernetes/features/issues/18

**Comment by bgrant0607** [comment]
The existing kube-up add-on manager has the additional problem that it doesn't use leader election. #46073

**Comment by bgrant0607** [comment]
More kinds of "add ons":
* Extension apiservers (API aggregation)
* Admission control hooks
* Authorization hook

**Comment by neolit123** [comment]
"/cluster/addons" been deprecated for a while.
the more recent pattern of applying addons is via operators or applying an operator that manages multiple addons:
related KEP:
https://github.com/kubernetes/enhancements/pull/746

the topic of generic "addon manager" that manages installation and lifecycle of addons is complicated and this is yet to be designed in a good state. in any case, it feels right to track this work under the cluster-addons subproject repository of sig-cluster-lifecycle:
https://github.com/kubernetes-sigs/cluster-addons
it already contains an addons installer idea:
https://github.com/kubernetes-sigs/cluster-addons/tree/master/installer
but projects like kubeadm and kops leaned towards managing addon operators directly and the above installer was considered more of a "standalone" tool.

given this ticket is tagged under cluster-lifecycle i'm going to close it, but feel free to log a new ticket under https://github.com/kubernetes-sigs/cluster-addons with the relevant ideas from this one.

thank you
/close
